ITE/10/README.adoc

```diff
@@ -0,0 +1,278 @@
+fit the materials and products model. This is a factor in _verifying_ such
+predicates using in-toto layouts.
+
+Therefore, all predicates supported for verification in an in-toto layout must
+categorized as one of two classes:
+
+* transformational predicates: these are based on classic in-toto links for
+  steps that transform artifacts in some manner. They consume some artifacts as
+  materials and produce others as products. The subject of such attestations
+  indicate the products of the step. Eg. in-toto links, SLSA Provenance.
+* informational predicates: these predicates record contextual attributes about
+  one or more artifacts. The operations they represent consume artifacts as
+  materials (recorded as the subject of such attestations) and produce
+  attributes about them as the outcome. Eg. test results, reviews.
+
+Note that transformational predicates may also record certain information about
+certain attributes. For example, the environment field in in-toto links can
+include such information.
+
+With this representation of different predicate types, step declarations in
+layouts can _chain_ different operations represented by distinct predicate types
+together, a fundamental feature of in-toto. For example, a code review step
+represented by a review predicate type that follows a checkout step that fetches
+source code represented by an in-toto link can use in-toto's artifact rules to
+ensure the review was performed against the sources fetched by the checkout
+step.
+
+=== Setting Predicate Type Expectations for Steps and Inspections
+
+Currently, step declarations have the following schema.
+
+```yaml
+name: string
+threshold: int
+expectedMaterials: list of artifact rules
+expectedProducts: list of artifact rules
+pubkeys: list of authorized key IDs
+expectedCommand: list of strings
+```
+
+This ITE proposes updating this schema to incorporate predicate types.
+
+```yaml
+name: string
+command: list of strings
+expectedMaterials: list of artifact rules
+expectedPredicates: list of expectedStepPredicates
+```
+
+Each `expectedStepPredicate` object has the following schema.
+
+```yaml
+predicateType: string
+expectedProducts: list of artifact rules
+functionaries: list of authorized functionary key IDs
+threshold: int
+```
+
+The `predicateType` field identifies the type of predicate expected for the
+step. `expectedProducts` is moved to being predicate specific as informational
+predicates may be generated alongside transformational predicates for the same
+step. This means that product rules do not generalize across all of the step's
+predicates. However, `expectedMaterials` is retained at the step level as all
+predicates generated for a step are expected to be for the same set of
+artifacts. `pubkeys` is updated to `functionaries` to generalize signing
+semantics. Finally, `threshold` applies per predicate type rather than the step
+as a whole.
+
+Similarly, inspections are also updated to support generating different
```

Comment:
Perhaps I'm obtuse but is `expectedPredicates` the list of predicates that are generated?  Reading the bit above I read this as the list of predicates that need to be presented by this step.

Reply:
Yes, `expectedPredicates` is what each step expects during verification. For inspections, this is just `predicates` because these attestations are generated at verification time (unsigned). I do wonder if we want to roll this back to what's in in-toto v1, where inspections generate links. If we allow multiple predicate types, the command executed during the inspection must be capable of generating an attestation of the specified type. This doesn't necessarily play well with the use cases we currently consider for inspections such as extracting from an archive to verify its contents etc. If we just stick to links, the verifier can continue generating them like in-toto v1 already does.

IMO, the _result_ of verification (VSA, summary link) is separate from inspections. We want an attestation for successful verification whether there's an inspection or not.

Reply:
I think the issue is primarily with the workflow which I'll update in the coming days. But more generally, I think this speaks to identifying whether inspections should generate custom predicates or not, period. The more I think about it, the more I wonder if we should leave inspections alone and just have them generate links, so no changes from in-toto v1.0 Let me take a pass at this.

Reply:
I think I'm in agreement, leaving inspections unchanged in this ITE makes a lot of sense to me.

Reply:
Updated to not discuss changes to inspections.